Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+
+            if (format.fourcc != preferredDMABufFormat) {
+                if (dumpBufferFormatNegotiation)
+                    LOG(Images, " -> incompatible format, does not match our preference, skipping.");
+                continue;
+            }
+
+            bool hasLinearModifier = false;
+            bool hasVivanteSuperTiledModifier = false;
+            if (useDRMModifiers) {
+                for (const auto& modifier : format.modifiers) {
+                    if (modifier == DRM_FORMAT_MOD_LINEAR)
+                        hasLinearModifier = true;
+                    else if (modifier == DRM_FORMAT_MOD_VIVANTE_SUPER_TILED)
+                        hasVivanteSuperTiledModifier = true;
+                }
+            }
+
+            useFormat = format;
+
+            if (useDRMModifiers) {
+                if (hasVivanteSuperTiledModifier && !shouldPreferDRMLinearModifier())
+                    useFormat.modifiers = { DRM_FORMAT_MOD_VIVANTE_SUPER_TILED };
+                else if (hasLinearModifier)
+                    useFormat.modifiers = { DRM_FORMAT_MOD_LINEAR };
+                else
+                    useFormat.modifiers = { };
+            } else
+                useFormat.modifiers = { };
+
+            break;
+        }
+
+        if (!useFormat.fourcc)
+            return std::nullopt;
+
+        if (dumpBufferFormatNegotiation) {
+            LOG(Images, " -> selected format, fourcc=0x%2x, n_modifiers=%zu, use modifiers? %i", useFormat.fourcc, useFormat.modifiers.size(), useDRMModifiers);
+
+            for (unsigned modifierIndex = 0; modifierIndex < useFormat.modifiers.size(); ++modifierIndex)
+                LOG(Images, " -> modifier[%2u]=0x%02lx", modifierIndex, useFormat.modifiers[modifierIndex]);
+        }
+
+        return useFormat;
+    };
+
+    auto bufferFormat = negotiateBufferFormat();
+    if (!bufferFormat.has_value()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to negotiate buffer format");
+        return nullptr;
+    }
+
+    auto gbmBuffer = makeUnique<MemoryMappedGPUBuffer>(size);
+    if (!gbmBuffer->allocate(device, bufferFormat.value())) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->map()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to map GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->createEGLImageFromDMABuf()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to export GBM buffer as EGLImage");
+        return nullptr;
+    }
+
+    return gbmBuffer;
+}
+
+bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
+{
+    m_modifier = DRM_FORMAT_MOD_INVALID;
+    if (bufferFormat.modifiers.size() > 0)
```

Comment:
```suggestion
    if (!bufferFormat.modifiers.isEmpty())
```

This is a personal preference, and might even be against "custom", but I tend to use empty checks when we don't care about the size

Reply:
That's right, preferred WebKit style too. Thanks.